_drafts/installing-gdb.md

```diff
@@ -0,0 +1,514 @@
+---
+title: "Tools we use: installing `gdb` for ARM"
+description: Demonstrate a few methods of installing gdb for ARM
+author: noah
+image: img/<post-slug>/cover.png # 1200x630
+---
+
+<!-- excerpt start -->
+
+In this mini article, I'll be going on a few strategies and nuances around
+getting a copy of GDB installed for debugging ARM chips.
+
+<!-- excerpt end -->
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+## GDB for ARM
+
+GDB, the GNU Project Debugger, is by default typically compiled to target the
+same architecture as the host system. When we talk about "host" and "target
+architectures, what we mean is:
+
+1. the **host** architecture: where the GDB program itself is run
+2. the **target** architecture: where the program being debugged is run
+
+Debugging a program that's running on the same machine as the host might look
+like this:
+
+[![img/installing-gdb/gdb_host_target.drawio.svg](/img/installing-gdb/gdb_host_target.drawio.svg)](/img/installing-gdb/gdb_host_target.drawio.svg)
+
+In embedded engineering, we often want to target a foreign architecture (eg, the
+embedded device, connected to some debug probe). That setup might look like
+this:
+
+[![img/installing-gdb/gdb_cross_target.drawio.svg](/img/installing-gdb/gdb_cross_target.drawio.svg)](/img/installing-gdb/gdb_cross_target.drawio.svg)
+
+We need a version of GDB that supports the target architecture of the program
+being debugged.
+
+## Summary of strategies
+
+I'm going to go through the various ways to install GDB with ARM support, but
+first here's a table summarizing the approaches:
+
+| Strategy                   | Binary or source install? | Pinned to version | Python support        |
+| -------------------------- | ------------------------- | ----------------- | --------------------- |
+| Build from source          | 📁source                  | ✅ yes            | ✅ (configurable)     |
+| Binaries from ARM          | 📦binary                  | ✅ yes            | ⚠️ varies per version |
+| System package manager     | 📦binary                  | ⚠️ varies         | ⚠️ varies             |
+| Conda                      | 📦binary                  | ✅ yes            | ✅ yes                |
+| Docker                     | 📦binary                  | ✅ yes            | ✅ yes                |
+| SDK-specific tools         | 📦binary                  | ⚠️ varies         | ✅ (usually yes)      |
+| xPack GNU Arm Embedded GCC | 📦binary                  | ✅ yes            | ✅ yes                |
+| crosstool-NG               | 📁source                  | ✅ yes            | ✅ (configurable)     |
+
+## Details on each strategy
+
+### Build from source
```

Comment:
I might put this one second-to-last. "When all else fails, you can always build from source!". And have Crosstool-NG last as "an easier way to go about this is..."

Reply:
ooo good call :+1: